native/src/main/java/io/ballerina/stdlib/nats/basic/consumer/DefaultMessageHandler.java

```diff
@@ -143,14 +129,29 @@ private void executeOnRequestResource(BMap<BString, Object>  msgObj, CountDownLa
                     NatsObservabilityConstants.CONTEXT_CONSUMER, connectedUrl,
                     msgObj.getStringValue(Constants.SUBJECT).getValue());
             properties.put(ObservabilityConstants.KEY_OBSERVER_CONTEXT, observerContext);
-            runtime.invokeMethodAsync(serviceObject, Constants.ON_REQUEST_RESOURCE, null, metadata,
-                                      new ResponseCallback(countDownLatch, subject, natsMetricsReporter, replyTo,
-                                                           this.natsConnection), properties,
-                    PredefinedTypes.TYPE_ANYDATA, msgObj, true);
+            if (serviceObject.getType().isIsolated() &&
+                    serviceObject.getType().isIsolated(Constants.ON_REQUEST_RESOURCE)) {
+                runtime.invokeMethodAsyncConcurrently(serviceObject, Constants.ON_REQUEST_RESOURCE, null, metadata,
+                        new ResponseCallback(countDownLatch, subject, natsMetricsReporter, replyTo,
+                                this.natsConnection), properties,
+                        PredefinedTypes.TYPE_ANYDATA, msgObj, true);
+            } else {
+                runtime.invokeMethodAsyncSequentially(serviceObject, Constants.ON_REQUEST_RESOURCE, null, metadata,
+                        new ResponseCallback(countDownLatch, subject, natsMetricsReporter, replyTo,
+                                this.natsConnection), properties,
+                        PredefinedTypes.TYPE_ANYDATA, msgObj, true);
+            }
```

Comment:
since this is kind of a common-case, shall we extract it out to a method ?

Reply:
Fixed in 302a65f73a7bf141a8517abceae525df86489c91